Add contributing.md file (#309)
scalacenter/scala-debug-adapter: CONTRIBUTING.md

- Java 8 or greater
- the sbt build tool
- one of the Scala code editor:
  - [VSCode](https://code.visualstudio.com/) with [Metals](https://marketplace.visualstudio.com/items?itemName=scalameta.metals) (recommended)
  - [IntelliJ IDEA Community Edition](https://www.jetbrains.com/idea/) with the [Scala plugin](https://plugins.jetbrains.com/plugin/1347-scala).

To install Java and sbt we recommend using Coursier (the official Scala installer) and following these [instructions](https://docs.scala-lang.org/getting-started/index.html).

## Setting Up

1. Clone the repository and its submodule

```
$ git clone git@github.com:scalacenter/scala-debug-adapter.git
$ cd scala-debug-adapter
$ git submodule update --init
```

2. Start the sbt shell in the terminal and compile

```
$ sbt
sbt:root> compile
```

3. Import the project in your code editor

Congratulations! You are now ready to start coding.

## Project Structure

All the modules are in the `modules` folder. They are composed of:
- `java-debug`: A fork of [microsoft/java-debug](https://github.com/microsoft/java-debug) on which the scala-debug-adapter depends.
The `java-debug` project was originally a Maven project.

> The `scala-debug-adapter` only depends on the `com.microsoft.java.debug.core` module and we compiles it directly using sbt.

Reviewer: ```suggestion
The `scala-debug-adapter` only depends on the `com.microsoft.java.debug.core` module and we compile it directly using sbt.
```

      StepIn.method("Predef$.println(Object)")
    )
  }
}
```

The organization of the tests is:
- `DebugServerTests`: general tests about starting and exiting debug sessions on all kinds of debugees
- `ScalaDebugTests`: general tests about inspecting threads, stack frames and variables in Scala programs
- `ScalaEvaluationTests`: tests about the Scala expression evaluator
- `SourceBreakpointTests`: tests about the different kind of source breakpoints: conditional breakpoints, hit count breakpoints and log points.
- `StepFilterTests`: tests about stepping in and out of Scala methods.

The tests are quite heavy to run because every test compiles a small Scala programs, creates a debug server that starts a fresh JVM to run the program, and executes a debugging scenario.
All of this is repeated on a few Scala versions.

Running all the test can take more than 30 minutes.
We recommend that locally you run only some tests using the sbt `testOnly` task or using your IDE:

```
sbt:root> testOnly ch.epfl.scala.debugadapter.Scala32EvaluationTests
```

## Running the debugger locally

You can install and start a local version of the scala-debug-adapter using Metals.

> 1. In sbt, check out the version of the scala-debug-adapter.

Reviewer: ```suggestion
1. In sbt, check out the version of the scala-debug-adapter project.
```

```
sbt:root> version
...
[info] 3.0.2-SNAPSHOT
```

In the following steps we will use `3.0.2-SNAPSHOT`.

2. Publish the scala-debug-adapter locally.

```
sbt:root> publishLocal
```

3. Open an sbt project in VS Code with Metals.
As we will run the debugger in this project, it must contain a main class or some test suites.

4. Open or create the `project/plugins.sbt` file.
This is where we can configure the version of the scala-debug-adapter that we want to use.

5. Configure the `sbt-debug-adapter` plugin in the `project/plugins.sbt` file.
Here you should use the local version that you just published, instead of `"3.0.2-SNAPSHOT"`.

```scala
addSbtPlugin("ch.epfl.scala" % "sbt-debug-adapter" % "3.0.2-SNAPSHOT")
```

6. Switch the build server. In VS Code, open the command palette, run `Metals: switch build server` and select `sbt`.

> 7. Since we changed the build file, we may need to reload sbt and reimport the build. You can do so by opening the command palette and run the `Metals: Import build` command, or by clicking on `Import build` in Metals'tab.

Reviewer: ```suggestion
7. Since we changed the build file, we may need to reload sbt and re-import the build. You can do so by opening the command palette and running the `Metals: Import build` command, or by clicking on `Import build` in Metals'tab.
```
